staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -510,7 +510,15 @@ func TestTransformingInformer(t *testing.T) {
 	podTransformer := func(obj interface{}) (interface{}, error) {
 		pod, ok := obj.(*v1.Pod)
 		if !ok {
-			return nil, fmt.Errorf("unexpected object type: %T", obj)
+			tombstone, isTombstone := obj.(DeletedFinalStateUnknown)
```

Comment:
I think we should handle tombstones prior to calling the transformer instead of making every transformer handle this case.

Reply:
+1

Reply:
Yup, agree, that seems like a very good idea! It does however seem harder and harder to avoid race conditions if mutating the input here, with all the edge cases... As noted by the test failure, when running this PR with `-race`, it pops up as a race condition.. :/

Eg. https://github.com/kubernetes/kubernetes/pull/116432/files#diff-53c0dd462f73f002a1153ee726720bdaed3ac218c5ebb82a691d352c0865209dR518 line leads to a race condition since the object is still in the cache when its handled as a tombstone (even without https://github.com/kubernetes/kubernetes/pull/115620). This becomes very tricky when dealing with transformer functions that mutate the input _and_ convert to another object type and doesn't return the input object/pointer...

From looking at the func definition, and the test example, it looks like we say mutating the input is ok; https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/client-go/tools/cache/controller.go#L356-L365

Reply:
Actually, if we do as @lavalamp suggests, we can skip the `TransformeFunc` inside `processDeltas` for tombstones, and then grab the latest properly converted object from the store, then give that to the user on `OnDelete` handle, inside a `DeletedFinalStateUnknown` object. That should keep the current behavior while getting rid of the races and the cumbersome transformers, right?

Reply:
no, the store's copy is potentially less recent than the copy in the deltas/tombstone.

I suggest a modification to transform objects prior to putting them in deltas at all, rather than when transferring from deltas to store. This will do extra work since not all deltas will make it all the way to the store, but it should be correct and not racy.

Alternatively we may need to store a bit to indicate if a thing has been transformed so that we only do that once; then we can do it lazily.

Reply:
> no, the store's copy is potentially less recent than the copy in the deltas/tombstone.

Hmm. If we pop the object from the store inside `processDeltas` when we hit a tombstone, shouldn't we then get the most recent version for the previous delta processed for that key? Or is there another edge case I'm missing?

> I suggest a modification to transform objects prior to putting them in deltas at all, rather than when transferring from deltas to store. This will do extra work since not all deltas will make it all the way to the store, but it should be correct and not racy.
>
> Alternatively we may need to store a bit to indicate if a thing has been transformed so that we only do that once; then we can do it lazily.

Yee, both those makes sense!

Reply:
A tombstone should be equally or more recent than whatever is in the store, we can't replace it with whatever is in the store, I think.

Reply:
> A tombstone should be equally or more recent than whatever is in the store, we can't replace it with whatever is in the store, I think.

Given we both talk about tombstones as `DeletedFinalStateUnknown` and _not_ all `DELETE` deltas, I think tombstones are with my changes always equal to the latest object in the store; since we grab the obj. from either the store or the newest delta that will be processed right before the tombstone delta. I can't see any cases where the tombstone object is newer than whats in the store.. Hmm

Reply:
Tho. I think your suggestion of running the conversion func before putting it in deltas is more elegant if we are able to do that.

Reply:
No, sometimes you produce a tombstone from a delta that hasn't been processed and placed in the store.

Reply:
> No, sometimes you produce a tombstone from a delta that hasn't been processed and placed in the store.

Yes, but when the tombstone delta gets to this line;
https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/tools/cache/controller.go#L441

all the previous deltas have been processed, and the newest possible object is present in the store.

Reply:
https://github.com/kubernetes/kubernetes/blob/48c8183fc42f2e64ecfde0c9fa8b03f637392a10/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L397

Reply:
I will take a stab at this.

Reply:
Take a look at #116623 and tell me if it improves things? if so we can combine with the test you wrote here.

Reply:
Thanks!